libs/s25main/notifications/NotificationManager_impl.h

```diff
@@ -58,62 +40,94 @@ struct NotificationManager::NoteCallbackBase
-    return Subscription(subscriber, CallbackUnregistrar<T_Note>(*this));
+    auto* subscriber = new NoteCallback<T_Note>(std::move(callback));
+    noteId2Subscriber[T_Note::getNoteId()].callbacks.push_back(subscriber);
+
+    return Subscription(subscriber, [this](void* subscription) {
+        RTTR_Assert(subscription); // As we use this in a shared_ptr, this can never be nullptr
+        auto* callback = static_cast<NoteCallback<T_Note>*>(subscription);
+        if(callback->IsSubscribed())
+            this->unsubscribe<T_Note>(callback);
+        delete callback;
+    });
 }
 
 template<class T_Note>
-void NotificationManager::unsubscribe(NoteCallback<T_Note>* callback)
+void NotificationManager::unsubscribe(NoteCallback<T_Note>* callback) noexcept
 {
-    if(!callback->IsSubscribed())
-        return;
-    if(isPublishing)
-        throw std::runtime_error("Cannot unsubscribe during publishing of messages");
-    CallbackList& callbacks = noteId2Subscriber[T_Note::getNoteId()];
+    RTTR_Assert(callback->IsSubscribed());
+    Subscribers& subs = noteId2Subscriber[T_Note::getNoteId()];
+    CallbackList& callbacks = subs.callbacks;
     auto itEl = std::find(callbacks.begin(), callbacks.end(), callback);
-    if(itEl != callbacks.end())
-        callbacks.erase(itEl);
+    RTTR_Assert(itEl != callbacks.end());
+    // We can't modify the list while iterating over it
+    if(subs.isPublishing)
+        *itEl = nullptr;
     else
-        RTTR_Assert(false);
+        callbacks.erase(itEl);
     // Set only the flag. Actual deletion of the callback must be handled by the shared_ptr deleter
     callback->SetUnsubscribed();
 }
 
 template<class T_Note>
 void NotificationManager::publish(const T_Note& notification)
 {
-    isPublishing = true;
-    for(auto* it : noteId2Subscriber[T_Note::getNoteId()])
-        static_cast<NoteCallback<T_Note>*>(it)->execute(notification);
-    isPublishing = false;
+    Subscribers& subs = noteId2Subscriber[T_Note::getNoteId()];
+    ++subs.isPublishing;
+    try
+    {
+        // Note: We have to support subscribe and unsubscribe during the execute call.
+        // - noteId2Subscriber must have elements with stable adresses even on insertion (subscribe)
+        // - CallbackList must not invalidate any references on insert
+        // - Since erase usually invalidates iterators unsubscribe only clears the pointer which we have to handle here
+        CallbackList& callbacks = subs.callbacks;
+        bool hasEmpty = false;
+        for(unsigned i = 0; i < callbacks.size(); ++i) // NOLINT(modernize-loop-convert)
+        {
+            auto* cb = static_cast<NoteCallback<T_Note>*>(callbacks[i]);
+            if(cb)
+                cb->execute(notification);
+            else
+                hasEmpty = true;
+        }
+        RTTR_Assert(subs.isPublishing);
+        --subs.isPublishing;
+        if(!subs.isPublishing && hasEmpty)
+            helpers::erase(callbacks, nullptr);
+    } catch(...)
```

Comment:
can you add a test which throws an exception inside execute? than this code should be triggered

Reply:
Good catch! Done.

Reply:
yes coveralls or so catched that :-D